Fix Runkit tags not being activated when comment is added

ledestin opened the pull request:
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

Fix Runkit tags not being activated when comment is added.

* Runkit tag activation was ran once, on page load. I've changed it to
  run on on comment preview and submit.
* It was necessary to add a check to skip already activated Runkit
  tags. The code didn't take into account that a tag could be already
  processed, and would just crash.

I feel that a better solution would be to add a listener on Runkit tag addition to DOM, and run tag activation, when that triggers. However, I'm not sure what browser support dev.to wants, so I decided to avoid that.

Haven't added any tests, because I'd like to get feedback first.

## Related Tickets & Documents

Fixes #1163

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)

## Added tests?

- [ ] yes
- [ ] no, because they aren't needed
- [x] no, because I need help

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/thepracticaldev/dev.to?pullRequest=6767) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/thepracticaldev/dev.to?pullRequest=6767) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/thepracticaldev/dev.to?pullRequest=6767) it.</sub>

ledestin commented:
Not sure what to do about Codeclimate here. The function is defined in Ruby and used in JS, and I guess that's why Codeclimate complains.

rhymes commented:
Hi @ledestin, thanks a lot for tackling this! 

I don't think it's working though, or at least now it breaks even the regular runkit tag.

This is production:

![runkit_prod](https://user-images.githubusercontent.com/146201/77253989-64145080-6c5e-11ea-8350-4ef04f2bd4cd.gif)

I tried with a simple runkit (with no comments), it takes a bit to load but it eventually does and functions.

This is this branch:

![runkit_dev_no_comment](https://user-images.githubusercontent.com/146201/77254002-7b533e00-6c5e-11ea-964f-1fc7f969e72e.gif)

Same code, nothing happens, but no error is logged in the console either. I am using Firefox 75.

I tried both with:

```liquid
{% runkit %}
console.log("hello, world");
{% endrunkit %}
```

and

```liquid
{% runkit 
// with hidden comment
%}
console.log("hello, world");
{% endrunkit %}
```

> However, I'm not sure what browser support dev.to wants, so I decided to avoid that.

We don't have an official list yet but basically Chromium based browsers (including the new Edge), Firefox and Safari (both mobile and desktop for all).

ledestin commented:
Hi @rhymes, thanks for testing. I've fixed the issue and added tests.

What should I do about CodeClimate? The coverage might be fixed by syncing the branch with upstream, but what should I do about the smell?

ledestin commented:
@rhymes To fix CodeClimate I'd need to rip out code from `RunkitTag` and put it in `app/assets/javascripts/initializers` perhaps, and I'm not sure it's such a good idea.

rhymes commented:
@ledestin yeah, I agree. You can signal the existence of `waitForRunkitAndActivateTags` using `global` like this:

```js
'use strict';

/* global  waitForRunkitAndActivateTags */

// the rest of the code
```

Please mark it ready for review whenever it is :)

ledestin commented:
@rhymes thanks! Hadn't realise it was ESLint complaining.

rhymes commented on app/views/articles/show.html.erb:
```diff
@@ -66,9 +66,7 @@
-  <% if @article.processed_html.include? "runkit-element" %>
-    <%= javascript_include_tag "https://embed.runkit.com" %>
-  <% end %>
+  <%= javascript_include_tag "https://embed.runkit.com" %>
```
is it possible to keep the if or with the new structure it isn't?

ledestin replied:
With the `if` present, **preview of the single new Runkit comment won't work**. The reason for that is probably that the initial show page without Runkit comments won't have Runkit lib, and the lib won't get loaded, when the "PREVIEW" button is clicked.

If the Runkit lib was loaded on demand, when clicking "PREVIEW", it'd be possible to keep the `if`, but I feel it'd get too complicated, with 2 places handling that.

ledestin commented:
@rhymes it takes 1-2 seconds for a comment preview for me to appear. I didn't change that much of the old code, just refactored and added additional Runkit tag activation on preview, etc.

This project is quite friendly :)

ledestin commented:
@rhymes um, I've added `defer`. For me,  https://embed.runkit.com/ load times are always in milliseconds, it might be the CDN server close to you is slow or the connection issues. With `traceroute`, you should be able to see, whether the problem is in the network path.

I thought more about it, and there's an option of dynamically loading  https://embed.runkit.com/. If we remove 2 places, where Rails includes it, and add calls to `waitForRunkitAndActivateTags()` on article preview, and possibly submit, that should work. Do you think it's worth doing it?

rhymes commented:
@ledestin it might be, it's around 100 kb uncompressed so it'd be awesome if there was a way to embed it only if needed

ledestin commented:
@rhymes switched to dynamic loading. Dried up some React code as I went.

nickytonline commented on app/views/liquids/_runkit.html.erb:
```diff
@@ -2,7 +2,7 @@
     <%= preamble %>
   </code>
-  <code>
+  <code style="display: none;">
```
@ledestin, I added this to avoid showing the code block before the iframe loads.

![May-11-2020 11-53-00](https://user-images.githubusercontent.com/833231/81582399-03f87b80-937e-11ea-8ce2-abb799f5ec84.gif)

nickytonline commented:
@ledestin, I merged the latest of master in your branch since this PR is almost a couple of months old.

nickytonline commented:
@ledesting, please grab latest of your branch as there is a spec failing for RunKit.

```bash
  1) RunkitTag#render generates proper div with content
     Failure/Error: Approvals.verify(rendered, name: "runkit_liquid_tag_spec", format: :html)
     
     Approvals::ApprovalError:
       Approval Error: Received file "/home/travis/build/thepracticaldev/dev.to/spec/support/fixtures/approvals/runkit_liquid_tag_spec.received.html" does not match approved "/home/travis/build/thepracticaldev/dev.to/spec/support/fixtures/approvals/runkit_liquid_tag_spec.approved.html".
     # ./spec/liquid_tags/runkit_tag_spec.rb:27:in `block (3 levels) in <top (required)>'
     # ./spec/rails_helper.rb:116:in `block (3 levels) in <top (required)>'
     # ./spec/rails_helper.rb:116:in `block (2 levels) in <top (required)>'
Finished in 15 minutes 5 seconds (files took 8.23 seconds to load)
4724 examples, 1 failure, 3 pending
Failed examples:
rspec ./spec/liquid_tags/runkit_tag_spec.rb:25 # RunkitTag#render generates proper div with content
Coverage report generated for RSpec to /home/travis/build/thepracticaldev/dev.to/coverage. 12645 / 13356 LOC (94.68%) covered.
The command "bundle exec rspec spec --color --tty" exited with 1.
```

ledestin commented:
@nickytonline good idea, thanks. Tests pass now.

rhymes commented on app/views/articles/_v2_form.html.erb:
```diff
@@ -51,6 +49,12 @@
   </div>
 </div>
 
+<% cache("article-edit-scripts", expires_in: 8.hours) do %>
```
I'm not sure we need to cache this right, there's nothing complicated to compute here, it's just the script content. 

If we decide to keep the cache, could we use `cache(cache_key_heroku_slug("article-v2-form-scripts"), expires_in: 8.hours)` here, this way it expires at each deploy

ledestin replied:
Good catch!

Well, the only expense of it is `.html_safe`, it seems. I've looked into it, and  `show.html.erb` caching was there from the start, but none of the tags currently use any computing.
https://github.com/thepracticaldev/dev.to/blame/583160532c694a14a20bf8e99ac1546258181352/app/views/articles/show.html.erb#L232

I'm going to remove the caching here then.

rhymes commented on app/views/articles/_v2_form.html.erb:
```diff
@@ -51,6 +49,12 @@
 
+<% cache("article-edit-scripts", expires_in: 8.hours) do %>
+  <script async>
+    <%= RunkitTag.script.html_safe %>
```
could you replace this with `<%== RunkitTag.script %>`.

The double `==` means it's not going to be escaped, similar to what `.html_safe` does

nickytonline commented on app/assets/javascripts/utilities/dynamicallyLoadScript.js:
```diff
@@ -0,0 +1,10 @@
+function dynamicallyLoadScript(url) {
+  if (document.querySelector(`script[src='${url}']`)) return;
+
+  var script = document.createElement('script');
```
```suggestion
  const script = document.createElement('script');
```
